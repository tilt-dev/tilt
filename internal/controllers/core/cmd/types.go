package cmd

import "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"

type Cmd = v1alpha1.Cmd
type CmdSpec = v1alpha1.CmdSpec
type CmdStatus = v1alpha1.CmdStatus
type CmdStateRunning = v1alpha1.CmdStateRunning
type CmdStateTerminated = v1alpha1.CmdStateTerminated

const LabelManifest = v1alpha1.LabelManifest
