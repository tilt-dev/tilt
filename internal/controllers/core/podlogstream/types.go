package podlogstream

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type PodLogStreamStatus = v1alpha1.PodLogStreamStatus
type ContainerLogStreamStatus = v1alpha1.ContainerLogStreamStatus
type ObjectMeta = metav1.ObjectMeta
