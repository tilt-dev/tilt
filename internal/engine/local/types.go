package local

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type Cmd = v1alpha1.Cmd
type CmdList = v1alpha1.CmdList
type CmdStatus = v1alpha1.CmdStatus
type CmdSpec = v1alpha1.CmdSpec
type CmdStateWaiting = v1alpha1.CmdStateWaiting
type CmdStateTerminated = v1alpha1.CmdStateTerminated
type CmdStateRunning = v1alpha1.CmdStateRunning
type ObjectMeta = metav1.ObjectMeta
