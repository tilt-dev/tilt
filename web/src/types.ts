import { Alert } from "./alerts"
import { Facet } from "./facets"

export enum SocketState {
  Loading,
  Reconnecting,
  Closed,
  Active,
}

export enum ResourceView {
  Log,
  Alerts,
  Facets = 2,
}

export enum TriggerMode {
  TriggerModeAuto,
  TriggerModeManualAfterInitial,
  TriggerModeManualIncludingInitial,
}

// what is the status of the resource in the cluster
export enum RuntimeStatus {
  Ok = "ok",
  Pending = "pending",
  Error = "error",
  NotApplicable = "not_applicable",
}

// What is the status of the resource with respect to Tilt
export enum ResourceStatus {
  Building, // Tilt is actively doing something (e.g., docker build or kubectl apply)
  Pending, // not building, healthy, or unhealthy, but presumably on its way to one of those (e.g., queued to build, or ContainerCreating)
  Healthy, // e.g., build succeeded and pod is running and healthy
  Unhealthy, // e.g., last build failed, or CrashLoopBackOff
  Warning, // e.g., an undismissed restart
  None, // e.g., a manual build that has never executed
}

export type SnapshotHighlight = {
  beginningLogID: string
  endingLogID: string
  text: string
}

export enum ShowFatalErrorModal {
  Default,
  Show,
  Hide,
}

export type Snapshot = {
  view: Proto.webviewView
  isSidebarClosed: boolean
  path?: string
  snapshotHighlight?: SnapshotHighlight | null
}

// A plaintext representation of a line of the log,
// with metadata to render it in isolation.
//
// The metadata should be stored as primitive fields
// so that React's default caching behavior will kick in.
export type LogLine = {
  // We assume that 'text' does not contain a newline
  text: string
  manifestName: string
  level: string
}
