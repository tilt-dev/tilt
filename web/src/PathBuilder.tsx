// A little helper class for building paths relative to the root of the app.

class PathBuilder {
  private host: string
  private snapId: string = ""

  constructor(host: string, pathname: string) {
    this.host = host

    const snapshotRe = new RegExp("^/snapshot/([^/]+)")
    let snapMatch = snapshotRe.exec(pathname)
    if (snapMatch) {
      this.snapId = snapMatch[1]
    }
  }

  getDataUrl() {
    if (this.isSnapshot()) {
      return this.snapshotDataUrl()
    }
    return `ws://${this.host}/ws/view`
  }

  isSnapshot(): boolean {
    return this.snapId !== ""
  }

  private snapshotDataUrl(): string {
    return `/api/snapshot/${this.snapId}`
  }

  private snapshotPathBase(): string {
    return `/snapshot/${this.snapId}`
  }

  rootPath() {
    if (this.isSnapshot()) {
      return this.snapshotPathBase()
    }
    return ""
  }

  path(relPath: string) {
    if (relPath[0] !== "/") {
      throw new Error('relPath should start with "/", actual:' + relPath)
    }
    return this.rootPath() + relPath
  }
}

export default PathBuilder
