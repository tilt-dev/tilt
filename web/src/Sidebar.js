import React, { PureComponent } from 'react';
import './Sidebar.scss';

class Sidebar extends PureComponent {
  render() {
    let classes = ['Sidebar']
    if (this.props.isOpen) {
      classes.push('is-open')
    }
    return (<div className={classes.join(' ')}>I'm a sidebar!</div>)
  }
}

export default Sidebar;
