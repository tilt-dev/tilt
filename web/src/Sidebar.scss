@import "constants";
@import "animations";

.Sidebar {
  position: fixed;
  top: 0;
  right: 0;
  bottom: $statusbar-height;
  width: $sidebar-width;
  background-color: $color-gray-darkest;
  box-sizing: border-box;
  overflow-y: auto;
  transform: translateX(0%);
  transition: transform ease $animation-timing;
  font-size: $font-size;
  display: flex;
  flex-direction: column;
  z-index: $z-sidebar;
}
.Sidebar.is-closed {
  transform: translateX(calc(100% - #{$sidebar-collapsed-width}));
}

// Collapse/Expand
.Sidebar-toggle {
  background-color: $color-gray-darkest;
  border: 0 none;
  color: inherit;
  font-size: inherit;
  font-family: inherit;
  display: flex;
  align-items: center;
  min-height: $sidebar-item;
  margin: 0;
  padding: 0;
  @include button-text;
  cursor: pointer;
  transition-property: color, background-color;
  transition-duration: $animation-timing;
  transition-timing-function: ease;
}
.Sidebar-toggle:hover {
  color: $color-blue-light;
}
.Sidebar-toggle > svg {
  fill: $color-white;
  width: $sidebar-collapsed-width;
  transition: transform $animation-timing ease-in, fill $animation-timing ease;
}
.Sidebar-toggle:hover > svg {
  fill: $color-blue-light;
}
.Sidebar.is-closed .Sidebar-toggle > svg {
  transform: rotate(180deg);
}
