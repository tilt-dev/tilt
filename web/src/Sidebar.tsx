import React, { PureComponent } from "react"
import { ReactComponent as ChevronSvg } from "./assets/svg/chevron.svg"
import { Link } from "react-router-dom"
import { combinedStatus } from "./status"
import "./Sidebar.scss"
import { ResourceStatus, ResourceView, TriggerMode } from "./types"
import TimeAgo from "react-timeago"
import { timeAgoFormatter } from "./timeFormatters"
import { isZeroTime } from "./time"
import PathBuilder from "./PathBuilder"
import SidebarIcon from "./SidebarIcon"
import SidebarTriggerButton from "./SidebarTriggerButton"
import { numberOfAlerts } from "./alerts"
import styled, { keyframes } from "styled-components"
import {
  AnimDuration,
  Color,
  ColorAlpha,
  ColorRGBA,
  FontSize,
  Height,
  SizeUnit,
  Width,
} from "./style-helpers"

const moment = require("moment")

type Resource = Proto.webviewResource
type Build = Proto.webviewBuildRecord

class SidebarItem {
  name: string
  isTiltfile: boolean
  status: ResourceStatus
  hasEndpoints: boolean
  lastBuildDur: Date | null
  lastDeployTime: string
  pendingBuildSince: string
  currentBuildStartTime: string
  alertCount: number
  triggerMode: TriggerMode
  hasPendingChanges: boolean
  queued: boolean
  lastBuild: Build | null = null

  /**
   * Create a pared down SidebarItem from a ResourceView
   */
  constructor(res: Resource) {
    let buildHistory = res.buildHistory || []
    let lastBuild = buildHistory.length > 0 ? buildHistory[0] : null

    this.name = res.name ?? ""
    this.isTiltfile = !!res.isTiltfile
    this.status = combinedStatus(res)
    this.hasEndpoints = (res.endpoints || []).length > 0
    this.lastBuildDur =
      lastBuild && lastBuild.startTime && lastBuild.finishTime
        ? timeDiff(lastBuild.startTime, lastBuild.finishTime)
        : null
    this.lastDeployTime = res.lastDeployTime ?? ""
    this.pendingBuildSince = res.pendingBuildSince ?? ""
    this.currentBuildStartTime = res.currentBuild?.startTime ?? ""
    this.alertCount = numberOfAlerts(res)
    this.triggerMode = res.triggerMode ?? TriggerMode.TriggerModeAuto
    this.hasPendingChanges = !!res.hasPendingChanges
    this.queued = !!res.queued
    this.lastBuild = lastBuild
  }
}

const barberpole = keyframes`
  100% {
    background-position: 100% 100%;
  }
`

let SidebarResources = styled.nav`
  flex: 1 0 auto;
  margin-left: ${SizeUnit(0.2)};
  margin-right: ${SizeUnit(0.2)};
`

let SidebarList = styled.ul`
  list-style: none;
`

let SidebarItemStyle = styled.li`
  width: 100%;
  color: ${Color.white};
  background-color: ${Color.gray};
  display: flex;
  align-items: stretch;
  height: ${Height.sidebarItem}px;
  transition: color ${AnimDuration.default} linear,
    background-color ${AnimDuration.default} linear;
  border-radius: ${SizeUnit(0.15)};
  overflow: hidden;
  position: relative; // Anchor the .isBuilding::after psuedo-element

  & + & {
    margin-top: ${SizeUnit(0.2)};
  }

  &:hover {
    background-color: ${ColorRGBA(Color.gray, ColorAlpha.translucent)};
    color: ${Color.blue};
  }

  &.isSelected {
    background-color: ${Color.white};
    color: ${Color.gray};
  }

  &.isBuilding::after {
    content: "";
    position: absolute;
    pointer-events: none;
    width: 100%;
    top: 0;
    bottom: 0;
    background: repeating-linear-gradient(
      225deg,
      ${ColorRGBA(Color.grayLight, ColorAlpha.translucent)},
      ${ColorRGBA(Color.grayLight, ColorAlpha.translucent)} 1px,
      ${ColorRGBA(Color.black, 0)} 1px,
      ${ColorRGBA(Color.black, 0)} 6px
    );
    background-size: 200% 200%;
    animation: ${barberpole} 8s linear infinite;
  }
`

let SidebarItemLink = styled(Link)`
  display: flex;
  align-items: stretch;
  text-decoration: none;
  // To truncate long names, root element needs an explicit width (i.e., not flex: 1)
  width: calc(100% - ${Width.sidebarTriggerButton}px);
`
let SidebarItemAll = styled(SidebarItemStyle)`
  text-transform: uppercase;
  margin-top: ${SizeUnit(0.5)};
  margin-bottom: ${SizeUnit(0.2)};
`
let SidebarItemName = styled.p`
  color: inherit;
  display: flex;
  align-items: center;
  flex: 1;
  overflow: hidden; // Reinforce truncation
`
// This child element helps truncated names show ellipses properly:
let SidebarItemNameTruncate = styled.span`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`
let SidebarTiming = styled.div`
  font-size: ${FontSize.small};
  line-height: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-end;
  flex-basis: auto;
`
let SidebarItemDuration = styled.span`
  opacity: ${ColorAlpha.almostOpaque};
  color: inherit;
`
let SidebarItemTimeAgo = styled.span`
  color: inherit;
`

type SidebarProps = {
  isClosed: boolean
  items: SidebarItem[]
  selected: string
  toggleSidebar: any
  resourceView: ResourceView
  pathBuilder: PathBuilder
}

class Sidebar extends PureComponent<SidebarProps> {
  render() {
    let pb = this.props.pathBuilder
    let classes = ["Sidebar"]
    if (this.props.isClosed) {
      classes.push("is-closed")
    }

    let allLink =
      this.props.resourceView === ResourceView.Alerts
        ? pb.path("/alerts")
        : pb.path("/")
    let totalAlerts = this.props.items
      .map(i => i.alertCount)
      .reduce((sum, current) => sum + current, 0)

    let nothingSelected = !this.props.selected
    let allItem = (
      <SidebarItemAll className={nothingSelected ? "isSelected" : ""}>
        <SidebarItemLink to={allLink}>
          <SidebarIcon status={ResourceStatus.None} alertCount={totalAlerts} />
          <SidebarItemName>All</SidebarItemName>
        </SidebarItemLink>
      </SidebarItemAll>
    )

    let listItems = this.props.items.map(item => {
      let link = `/r/${item.name}`
      switch (this.props.resourceView) {
        case ResourceView.Alerts:
          link += "/alerts"
          break
        case ResourceView.Facets:
          link += "/facets"
          break
      }

      let formatter = timeAgoFormatter
      let hasSuccessfullyDeployed = !isZeroTime(item.lastDeployTime)
      let hasBuilt = item.lastBuild !== null
      let building = !isZeroTime(item.currentBuildStartTime)
      let buildDur = item.lastBuildDur ? formatDuration(item.lastBuildDur) : ""
      let timeAgo = <TimeAgo date={item.lastDeployTime} formatter={formatter} />
      let isSelected = this.props.selected === item.name

      let classes = ""
      if (building) {
        classes += " isBuilding"
      }
      if (isSelected) {
        classes += " isSelected"
      }
      return (
        <SidebarItemStyle key={item.name} className={classes}>
          <SidebarItemLink
            className="SidebarItem-link"
            to={pb.path(link)}
            title={item.name}
          >
            <SidebarIcon status={item.status} alertCount={item.alertCount} />
            <SidebarItemName>
              <SidebarItemNameTruncate>{item.name}</SidebarItemNameTruncate>
            </SidebarItemName>
            <SidebarTiming>
              <SidebarItemTimeAgo
                className={hasSuccessfullyDeployed ? "" : "isEmpty"}
              >
                {hasSuccessfullyDeployed ? timeAgo : "—"}
              </SidebarItemTimeAgo>
              <SidebarItemDuration
                className={hasSuccessfullyDeployed ? "" : "isEmpty"}
              >
                {hasSuccessfullyDeployed ? buildDur : "—"}
              </SidebarItemDuration>
            </SidebarTiming>
          </SidebarItemLink>
          <SidebarTriggerButton
            resourceName={item.name}
            isTiltfile={item.isTiltfile}
            isSelected={isSelected}
            hasPendingChanges={item.hasPendingChanges}
            hasBuilt={hasBuilt}
            isBuilding={building}
            triggerMode={item.triggerMode}
            isQueued={item.queued}
          />
        </SidebarItemStyle>
      )
    })

    return (
      <section className={classes.join(" ")}>
        <SidebarResources className="Sidebar-resources">
          <SidebarList>
            {allItem}
            {listItems}
          </SidebarList>
        </SidebarResources>
        <button className="Sidebar-toggle" onClick={this.props.toggleSidebar}>
          <ChevronSvg /> Collapse
        </button>
      </section>
    )
  }
}

function timeDiff(start: string, end: string): Date {
  let t1 = moment(start)
  let t2 = moment(end)
  return t2.diff(t1)
}

function formatDuration(dur: Date): string {
  // TODO(han) — hacky for now. The duration should count up during building!
  // We also need to display durations of 1m+ (ex: mm:ss)
  return moment(dur).format("s.S") + "s"
}

export default Sidebar

export { SidebarItem }
