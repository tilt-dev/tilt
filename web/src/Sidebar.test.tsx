import React from "react"
import { mount } from "enzyme"
import renderer from "react-test-renderer"
import { MemoryRouter } from "react-router"
import Sidebar, { SidebarItem } from "./Sidebar"
import { oneResource, twoResourceView } from "./testdata"
import { ResourceView, TriggerMode } from "./types"
import PathBuilder from "./PathBuilder"

type Resource = Proto.webviewResource

let pathBuilder = new PathBuilder("localhost", "/")

let realDateNow = Date.now

describe("sidebar", () => {
  beforeEach(() => {
    Date.now = jest.fn(() => Date.UTC(2017, 11, 21, 15, 36, 7, 0))
  })
  afterEach(() => {
    Date.now = realDateNow
  })
  it("renders empty resource list without crashing", () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <Sidebar
            isClosed={true}
            items={[]}
            selected=""
            toggleSidebar={null}
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders list of resources", () => {
    let items = twoResourceView().resources.map((res: Resource) => {
      let history = res.buildHistory ?? []
      history[0].error = "error!"
      return new SidebarItem(res)
    })
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <Sidebar
            isClosed={false}
            items={items}
            selected=""
            toggleSidebar={null}
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("abbreviates durations under a minute", () => {
    let items = [4, 9, 19, 29, 39, 49, 54].map(d => {
      let res = oneResource()
      res.name = `resource${d}`
      res.lastDeployTime = new Date(Date.now() - d * 1000).toISOString()
      return new SidebarItem(res)
    })

    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <Sidebar
            isClosed={false}
            items={items}
            selected=""
            toggleSidebar={null}
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders resources that haven't been built yet", () => {
    let items = twoResourceView().resources.map((res: any) => {
      // currently building, no completed builds
      res.lastDeployTime = "0001-01-01T00:00:00Z"
      return new SidebarItem(res)
    })
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <Sidebar
            isClosed={false}
            items={items}
            selected=""
            toggleSidebar={null}
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
