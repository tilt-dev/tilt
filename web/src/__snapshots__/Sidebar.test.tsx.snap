// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar abbreviates durations under a minute 1`] = `
<nav
  className="sc-AxhCb iWVPDl Sidebar-resources"
>
  <ul
    className="sc-AxhUy hFLisE"
  >
    <li
      className="sc-AxgMl sc-Axmtr IezSP isSelected"
    >
      <a
        className="sc-AxheI bUZbcE"
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-AxjAm hcburE isNone"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            7
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource4"
        onClick={[Function]}
        title="resource4"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource4
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:36:03.000Z"
              title="2017-12-21T15:36:03.000Z"
            >
              &lt;5s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource9"
        onClick={[Function]}
        title="resource9"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource9
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:58.000Z"
              title="2017-12-21T15:35:58.000Z"
            >
              &lt;15s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource19"
        onClick={[Function]}
        title="resource19"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource19
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:48.000Z"
              title="2017-12-21T15:35:48.000Z"
            >
              &lt;30s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource29"
        onClick={[Function]}
        title="resource29"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource29
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:38.000Z"
              title="2017-12-21T15:35:38.000Z"
            >
              &lt;30s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource39"
        onClick={[Function]}
        title="resource39"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource39
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:28.000Z"
              title="2017-12-21T15:35:28.000Z"
            >
              &lt;45s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource49"
        onClick={[Function]}
        title="resource49"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource49
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:18.000Z"
              title="2017-12-21T15:35:18.000Z"
            >
              &lt;1m ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/resource54"
        onClick={[Function]}
        title="resource54"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            resource54
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:13.000Z"
              title="2017-12-21T15:35:13.000Z"
            >
              &lt;1m ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
</nav>
`;

exports[`sidebar renders empty resource list without crashing 1`] = `
<nav
  className="sc-AxhCb iWVPDl Sidebar-resources"
>
  <ul
    className="sc-AxhUy hFLisE"
  >
    <li
      className="sc-AxgMl sc-Axmtr IezSP isSelected"
    >
      <a
        className="sc-AxheI bUZbcE"
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-AxjAm hcburE isNone"
        >
          <span>
             
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          All
        </p>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`sidebar renders list of resources 1`] = `
<nav
  className="sc-AxhCb iWVPDl Sidebar-resources"
>
  <ul
    className="sc-AxhUy hFLisE"
  >
    <li
      className="sc-AxgMl sc-Axmtr IezSP isSelected"
    >
      <a
        className="sc-AxheI bUZbcE"
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-AxjAm hcburE isNone"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            2
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/vigoda"
        onClick={[Function]}
        title="vigoda"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            vigoda
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:36:07.000Z"
              title="2017-12-21T15:36:07.000Z"
            >
              &lt;5s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/snack"
        onClick={[Function]}
        title="snack"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            snack
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa"
          >
            <time
              dateTime="2017-12-21T15:35:57.000Z"
              title="2017-12-21T15:35:57.000Z"
            >
              &lt;15s ago
            </time>
          </span>
          <span
            className="sc-fzpans clZIwH"
          >
            -10.0s
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
</nav>
`;

exports[`sidebar renders resources that haven't been built yet 1`] = `
<nav
  className="sc-AxhCb iWVPDl Sidebar-resources"
>
  <ul
    className="sc-AxhUy hFLisE"
  >
    <li
      className="sc-AxgMl sc-Axmtr IezSP isSelected"
    >
      <a
        className="sc-AxheI bUZbcE"
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-AxjAm hcburE isNone"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            2
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/vigoda"
        onClick={[Function]}
        title="vigoda"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            vigoda
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa isEmpty"
          >
            —
          </span>
          <span
            className="sc-fzpans clZIwH isEmpty"
          >
            —
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-AxgMl kiZofC  isBuilding"
    >
      <a
        className="sc-AxheI bUZbcE SidebarItem-link"
        href="/r/snack"
        onClick={[Function]}
        title="snack"
      >
        <div
          className="sc-AxjAm hcburE isBuilding"
        >
          <span
            className="sc-AxirZ feYmtM"
          >
            1
          </span>
        </div>
        <p
          className="sc-AxmLO bqHpRS"
        >
          <span
            className="sc-fzozJi jcfBR"
          >
            snack
          </span>
        </p>
        <div
          className="sc-fzoLsD fjbaZw"
        >
          <span
            className="sc-fzplWN jEKYIa isEmpty"
          >
            —
          </span>
          <span
            className="sc-fzpans clZIwH isEmpty"
          >
            —
          </span>
        </div>
      </a>
      <button
        className="sc-AxiKw gOpblw SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
</nav>
`;
